File: 872 - Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        l1, l2 = [], []
        l1 = self.getLeaf(root1, l1)
        l2 = self.getLeaf(root2, l2)
        return l1 == l2

    def getLeaf(self, root, leaves):
        if root is None:
            return leaves
        if root.left is not None:
            leaves = self.getLeaf(root.left, leaves)
        if root.right is not None:
            leaves = self.getLeaf(root.right, leaves)
        if root.left is None and root.right is None:
            leaves.append(root.val)
        return leaves

File: 872 - Leaf-Similar Trees.java
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> l1 = new ArrayList<>(), l2 = new ArrayList<>();
        l1 = getLeaves(root1, l1);
        l2 = getLeaves(root2, l2);
        return l1.equals(l2);
    }
    private List<Integer> getLeaves(TreeNode root, List<Integer> leaves) {
        if (root == null) return leaves;
        leaves = getLeaves(root.left, leaves);
        leaves = getLeaves(root.right, leaves);
        if (root.left == null && root.right == null) {
            leaves.add(root.val);
        }
        return leaves;
    }
}